devito/finite_differences/finite_difference.py

```diff
@@ -161,184 +163,60 @@ def second_derivative(expr, **kwargs):
-            subs = dict([(d, ind[i].subs({dim: d})) for d in all_dims])
-            deriv += coeffs[i] * expr.subs(subs)
-    return deriv.evalf(_PRECISION)
+    return generic_derivative(expr, dim, fd_order, 2, stagger=stagger)
 
 
 @check_input
-def generic_derivative(expr, deriv_order, dim, fd_order, **kwargs):
+def cross_derivative(expr, dims, fd_order, deriv_order, stagger=None):
     """
-    Arbitrary-order derivative of a given expression.
+    Arbitrary-order cross derivative of a given expression.
 
     Parameters
     ----------
     expr : expr-like
-        Expression for which the derivative is produced.
-    deriv_order : int
-        Derivative order, e.g. 2 for a second-order derivative.
-    dim : Dimension
-        The Dimension w.r.t. which to differentiate.
+        Expression for which the cross derivative is produced.
+    dims : tuple
+        Dimensions w.r.t. which to differentiate.
     fd_order : int
         Coefficient discretization order. Note: this impacts the width of
         the resulting stencil.
-
-    Returns
-    -------
-    expr-like
-        The derivative of ``expr`` of order ``deriv-order``.
-    """
-    indices = [(dim + i * dim.spacing) for i in range(-fd_order//2, fd_order//2 + 1)]
-    if fd_order == 1:
-        indices = [dim, dim + dim.spacing]
-    c = finite_diff_weights(deriv_order, indices, dim)[-1][-1]
-    deriv = 0
-    all_dims = tuple(set((dim, ) +
-                     tuple([i for i in expr.indices if i.root == dim])))
-    for i in range(0, len(indices)):
-            subs = dict([(d, indices[i].subs({dim: d})) for d in all_dims])
-            deriv += expr.subs(subs) * c[i]
-
-    return deriv.evalf(_PRECISION)
-
-
-@check_input
-def cross_derivative(expr, **kwargs):
-    """
-    Cross derivative of a given expression.
-
-    Parameters
-    ----------
-    expr : expr-like
-        Expression for which the cross derivative is produced.
-    **kwargs
-        - ``dims``: 2-tuple Dimensions w.r.t. which to differentiate.
-        - ``diff``: Finite-difference symbol, defaults to `h`.
-        - ``order``: 2-tuple representing the discretisation order of the
-                     derivative in the two Dimensions.
-
-    Returns
-    -------
-    expr-like
-        The cross derivative of ``expr``
+    deriv_order : int
+        Derivative order, e.g. 2 for a second-order derivative.
+    stagger : tuple
+        Staggering of the input expression in each dimension of the cross derivative
 
     Examples
     --------
-    >>> from devito import Function, Grid, cross_derivative
+    >>> from devito import Function, Grid, second_derivative
     >>> grid = Grid(shape=(4, 4))
     >>> x, y = grid.dimensions
     >>> f = Function(name='f', grid=grid, space_order=2)
     >>> g = Function(name='g', grid=grid, space_order=2)
-    >>> # cross_derivative(f*g, dims=(x, y))
-    TODO
+    >>> cross_derivative(f*g, dims=(x, y), fd_order=(2, 2), deriv_order=(1, 1))
```

Comment:
can we put something (much) simpler here? this is really verbose

Reply:
Not really no, it's a cross derivative so verbose anyway